### PR TITLE
fix: `istanbul-lib-source-maps` to preserve implicit `else` when sourcemaps are used

### LINKED ISSUE #705: `istanbul-lib-source-maps` does not preserve implicit `else`
When the coverage map is mapped back to sources using `istanbul-lib-source-maps` the implicit `else` gets dropped in the process. Here is the line that checks whether the branch exists in sources, and simply excludes it from results. Since the implicit `else` does not exist in the sources, the `mapping` variable is falsy.

https://github.com/istanbuljs/istanbuljs/blob/fb8cb4bb99c9e5d36c8b011338d2ab5d3c72ff9c/packages/istanbul-lib-source-maps/lib/transformer.js#L71-L85

Example of implicit `else` just so that everyone is on the track of what this issue is about:

```js
// Source code
if(condition) {
  method();
}
unrelatedMethod();

// Instrumented
if(condition) {
  coverageMap.b[0]++;
  method();
} else { // Implicit else right here!
  coverageMap.b[1]++;
}
unrelatedMethod();
```

Check the reproduction case below to see how `data.branchMap` and `data.b` loses the `else` branch.

<details>
  <summary>Minimal reproduction case</summary>

Requirements:
```sh
$ npm install istanbul-lib-instrument istanbul-lib-coverage istanbul-lib-source-maps esbuild
$ node -v
> v16.15.1
```

```js
const filename = "/math.ts";

function logBranches(label, coverageMap) {
  console.log(label);
  console.log(JSON.stringify(coverageMap.data[filename].branchMap, null, 2));
  console.log(coverageMap.data[filename].b);
}

async function run() {
  const sourceCode = `
function add(a: number, b: number): number {
    type SomeType = typeof a | string;

    console.log('Running add');

    if(a === 20) {
        return b;
    }
}

add(1, 2);
`;

  const transpiled = require("esbuild").transformSync(sourceCode, {
    loader: "ts",
    sourcemap: true,
    sourcefile: filename,
  });
  transpiled.map = JSON.parse(transpiled.map);

  const instrumenter = require("istanbul-lib-instrument").createInstrumenter({
    produceSourceMap: true,
    autoWrap: false,
    compact: false,
    coverageVariable: "__coverage__",
    coverageGlobalScope: "globalThis",
  });

  const instrumented = instrumenter.instrumentSync(
    transpiled.code,
    filename,
    transpiled.map
  );

  // Run instrumented code to populate globalThis.__coverage__
  eval(instrumented);
  const __coverage__ = globalThis.__coverage__;

  const { createCoverageMap } = require("istanbul-lib-coverage");
  const coverageMap = createCoverageMap(__coverage__);

  const { createSourceMapStore } = require("istanbul-lib-source-maps");
  const sourceMapStore = createSourceMapStore();
  const transformedCoverage = await sourceMapStore.transformCoverage(
    coverageMap
  );

  logBranches("coveragemap before transforming back to sources", coverageMap);
  logBranches("coveragemap after transforming to sources", transformedCoverage);
}

run();
```

Output:

```
$ node index.js 
Running add
coveragemap before transforming back to sources
{
  "0": {
    "loc": {
      "start": {
        "line": 3,
        "column": 2
      },
      "end": {
        "line": 5,
        "column": 3
      }
    },
    "type": "if",
    "locations": [
      {
        "start": {
          "line": 3,
          "column": 2
        },
        "end": {
          "line": 5,
          "column": 3
        }
      },
      {  // See here!
        "start": {},
        "end": {}
      }
    ],
    "line": 3
  }
}
{ '0': [ 0, 1 ] }  // See here!

coveragemap after transforming to sources
{
  "0": {
    "loc": {
      "start": {
        "line": 7,
        "column": 4
      },
      "end": {
        "line": 9,
        "column": null
      }
    },
    "type": "if",
    "locations": [
      {
        "start": {
          "line": 7,
          "column": 4
        },
        "end": {
          "line": 9,
          "column": null
        }
      }
       // See here!
    ]
  }
}
{ '0': [ 0 ] } // See here!
```

</details>

Related issues from downstream:
- https://github.com/vitest-dev/vitest/issues/1887
- https://github.com/vitest-dev/vitest/issues/2239

### PR DESCRIPTION
**Pull request by AriPerkkio**
- Fixes #705 

Tested the fix with reproduction case from downstream (https://github.com/vitest-dev/vitest/issues/2239) by modifying `./node_modules/istanbul-lib-source-maps/lib/transformer.js` on the fly. 

```
----------|---------|----------|---------|---------|-------------------
File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------|---------|----------|---------|---------|-------------------
All files |     100 |       50 |     100 |     100 |                   
 cov.js   |     100 |       50 |     100 |     100 | 2                 
----------|---------|----------|---------|---------|-------------------

```

### CODE REVIEW

File: packages/istanbul-lib-source-maps/lib/transformer.js [hunk 1]
```diff
@@ -82,6 +82,18 @@ class SourceMapTransformer {
...
+                    branchMeta.type === 'if' &&
+                    i > 0 &&
+                    loc.start.line === undefined &&
+                    loc.start.end === undefined &&
+                    loc.end.line === undefined &&
+                    loc.end.end === undefined
+                ) {
+                    locs.push(loc);
+                    mappedHits.push(hits[i]);
+                }
```
Comment (AriPerkkio) [inline packages/istanbul-lib-source-maps/lib/transformer.js:96]: For reference:

https://github.com/istanbuljs/istanbuljs/blob/fb8cb4bb99c9e5d36c8b011338d2ab5d3c72ff9c/docs/raw-output.md?plain=1#L49

https://github.com/istanbuljs/istanbuljs/blob/fb8cb4bb99c9e5d36c8b011338d2ab5d3c72ff9c/docs/raw-output.md?plain=1#L52

https://github.com/istanbuljs/istanbuljs/blob/fb8cb4bb99c9e5d36c8b011338d2ab5d3c72ff9c/packages/istanbul-reports/lib/html/annotator.js#L147-L154

**Comment by adrian-burlacu-software** [comment]
@AriPerkkio
Thank you, I appreciate this! Felt pretty embarrassed trying to find all the spots that needed to update downstream.

**Comment by carpie** [comment]
Hello! Is there a timeline for this to be merged? We're hitting this issue as well with cypress code coverage on instrumented code with source maps.  Thanks!

**Comment by AriPerkkio** [comment]
Istanbul projects are not that active anymore so updates may take long time - if ever. 

If you have access to `coverageMap` in your Cypress setup you can apply same work-around that Vitest uses to patch these implicit else's: https://github.com/vitest-dev/vitest/pull/2275/files#diff-a0a730c31ce78cf1c3cb20d55daa7167ff57fa8e062df213c6eeb918582d1cb5R244-R272

**Review by SimenB — CHANGES_REQUESTED** [review]
Since no old tests are changed, I assume this is fine 🙂

CI seems to be failing, tho

**Comment by SimenB** [comment]
Right, that makes sense 👍

**Review by SimenB — APPROVED** [review]

**Comment by SimenB** [comment]
https://github.com/istanbuljs/istanbuljs/releases/tag/istanbul-lib-source-maps-v5.0.5

**Comment by AriPerkkio** [comment]
This change is actually breaking some test cases in Vitest - tests that we didn't have in 2022 when this PR was created 😅 

`source` here can now be `null`:

https://github.com/istanbuljs/istanbuljs/blob/41d98a7c2a852a9cf8e49c087706bf7f71a91d4a/packages/istanbul-lib-source-maps/lib/transformer.js#L105

Which ends up breaking here:

https://github.com/istanbuljs/istanbuljs/blob/41d98a7c2a852a9cf8e49c087706bf7f71a91d4a/packages/istanbul-lib-source-maps/lib/transformer.js#L120-L127

https://github.com/istanbuljs/istanbuljs/blob/41d98a7c2a852a9cf8e49c087706bf7f71a91d4a/packages/istanbul-lib-source-maps/lib/transform-utils.js#L7-L9


https://github.com/vitest-dev/vitest/actions/runs/9743759050/job/26887930830?pr=6014

```
TypeError: Cannot read properties of undefined (reading 'replace')
 ❯ getUniqueKey ../../node_modules/.pnpm/istanbul-lib-source-maps@5.0.5/node_modules/istanbul-lib-source-maps/lib/transform-utils.js:8:21
 ❯ getMappedCoverage ../../node_modules/.pnpm/istanbul-lib-source-maps@5.0.5/node_modules/istanbul-lib-source-maps/lib/transformer.js:121:25
 ❯ ../../node_modules/.pnpm/istanbul-lib-source-maps@5.0.5/node_modules/istanbul-lib-source-maps/lib/transformer.js:105:40
 ❯ SourceMapTransformer.processFile ../../node_modules/.pnpm/istanbul-lib-source-maps@5.0.5/node_modules/istanbul-lib-source-maps/lib/transformer.js:64:38
 ❯ SourceMapTransformer.transform ../../node_modules/.pnpm/istanbul-lib-source-maps@5.0.5/node_modules/istanbul-lib-source-maps/lib/transformer.js:137:38
 ❯ MapStore.transformCoverage ../../node_modules/.pnpm/istanbul-lib-source-maps@5.0.5/node_modules/istanbul-lib-source-maps/lib/map-store.js:217:16
```

**Comment by AriPerkkio** [comment]
I can look for a fix tomorrow. Most of our test cases are passing just fine so it looks like some kind of edge case. Maybe most users won't run into it.

**Comment by SimenB** [comment]
Can see if we have any failing tests as well: https://github.com/jestjs/jest/pull/15156

**Comment by AriPerkkio** [comment]
Alright fix ready in https://github.com/istanbuljs/istanbuljs/pull/789. I've tested it with Vitest.
